### ISSUE TITLE
[AOTInductor] max_autotune non-deterministically selects cuBLAS over fused Triton kernels for small GEMM, causing significant performance regression

### ISSUE THREAD

**Issue by AllenFeiZZ** [pytorch/pytorch#180811]
## How to reproduce
The issue occurs when using AOTInductor with `max_autotune=True` on a model that has `nn.Linear` + activation (GeLU/ReLU) patterns with small batch (M=1). Clear the inductor cache and compile twice — the generated `wrapper.cpp` may non-deterministically contain `aoti_torch_cuda_addmm_out` (cuBLAS) calls vs. fused `triton_tem_fused_addmm_gelu_*` launches.

**Our scenario**: ~50 FFN layers doing GEMM with shape M=1, N=200, K=800 followed by GeLU. Compiled with:
```python
cfg.max_autotune = True
cfg.max_autotune_gemm_backends = "ATEN,CUTLASS,TRITON"  # default
```

**Results across compilations** (same model, same env, cleared cache):
- **Compilation A**: 0 `aoti_torch_cuda_addmm_out` → all fused into Triton templates → fast
- **Compilation B**: 50 `aoti_torch_cuda_addmm_out` → fusion broken → slow 

### Evidence from wrapper.cpp diff

Fast compilation (0 cuBLAS):
```cpp
// Single fused kernel: addmm + gelu
launchKernel(kernels_.triton_tem_fused_addmm_gelu_66, grid_0, grid_1, grid_2, 4, 24576, kernel_args_, stream_);
```

Slow compilation (cuBLAS fallback):
```cpp
// cuBLAS GEMM (unfused)
aoti_torch_cuda_addmm_out(tmp_tensor_16, arg106_, buf91, arg107_, 1, 1);
// Separate pointwise gelu
launchKernel(kernels_.triton_poi_fused_gelu_67, grid_0, grid_1, grid_2, 4, 0, kernel_args_, stream_);

### Versions

## Environment

- PyTorch version: 2.8.0 (with AOTInductor support)
- GPU: NVIDIA Ada Lovelace (compute capability 8.9, 92 SMs)
- Python: 3.12
- OS: Linux

cc @jerryzh168 @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @mruberry @kurtamohler @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @ydwu4 @desertfire @yushangdi @iupaikov-amd

**Comment by yushangdi** [comment]
can you try on latest pytorch release? I cannot reproduce this on the latest nightly using H100.

I do not have access to Ada Lovelace, so cannot check there, but please let us know if you're still having issues with the latest release. 

```
Setup                                                                                                 
  - Trunk torch 2.13.0a0+git855f08b (uv venv at /data/users/shangdiy/pytorch/.venv)
  - triton 3.7.0                                                                                        
  - H100 (SM 9.0); the issue was filed on an Ada SM 8.9 GPU       
                                                                                                        
  What the repro ran (agent_space/repro_180811.py)                                                      
  - 50 × nn.Linear(800→200) + GeLU, M=1                                                                 
  - torch.export + torch._inductor.aot_compile                                                          
  - max_autotune=True, max_autotune_gemm_backends="ATEN,CUTLASS,TRITON"                                 
  - Fresh TORCHINDUCTOR_CACHE_DIR / TRITON_CACHE_DIR per run, 8 runs                                    
                                                                                                        
  Result: deterministic across all 8 runs — every run produced {cuBLAS_addmm_out: 0, cuBLAS_mm_out: 50, 
  triton_tem_fused_addmm: 0, triton_poi_fused: 68}.                     
```

My repro script:
```
"""
Repro for https://github.com/pytorch/pytorch/issues/180811
AOTInductor max_autotune non-deterministically selects cuBLAS over fused Triton
kernels for small GEMM (M=1, N=200, K=800) followed by GeLU.

Run multiple compilations with cache cleared each time, then count
`aoti_torch_cuda_addmm_out` occurrences in the generated wrapper.cpp.
If deterministic, count should be the same across runs.
"""

import argparse
import os
import re
import shutil
import tempfile
import zipfile
from pathlib import Path

import torch
import torch.nn as nn
from torch._inductor import config as inductor_config


NUM_LAYERS = 50
M, K, N = 1, 800, 200


class FFN(nn.Module):
    def __init__(self, n_layers=NUM_LAYERS, in_features=K, out_features=N):
        super().__init__()
        # Each layer: Linear(K -> N) + GeLU, preserving K for chaining we need N==K;
        # the issue is about individual GEMMs being fused with activation, so use
        # a list of independent Linears that all consume the same input (like FFN sublayers).
        self.layers = nn.ModuleList(
            [nn.Linear(in_features, out_features) for _ in range(n_layers)]
        )

    def forward(self, x):
        outs = []
        for layer in self.layers:
            y = layer(x)
            y = torch.nn.functional.gelu(y)
            outs.append(y)
        return torch.stack(outs, dim=0)


def run_one(run_idx: int, workdir: Path) -> dict:
    """Compile once with a fresh cache via aoti_compile_and_package, then
    extract wrapper.cpp from the .pt2 package and return kernel counts."""

    # Fresh caches per run
    cache_dir = workdir / f"inductor_cache_{run_idx}"
    triton_cache_dir = workdir / f"triton_cache_{run_idx}"
    out_dir = workdir / f"out_{run_idx}"
    for d in (cache_dir, triton_cache_dir, out_dir):
        if d.exists():
            shutil.rmtree(d)
        d.mkdir(parents=True)

    os.environ["TORCHINDUCTOR_CACHE_DIR"] = str(cache_dir)
    os.environ["TRITON_CACHE_DIR"] = str(triton_cache_dir)

    torch.manual_seed(0)
    model = FFN().cuda().eval()
    example = (torch.randn(M, K, device="cuda"),)

    package_path = str(out_dir / "model.pt2")
    with inductor_config.patch(
        {
            "max_autotune": True,
            "max_autotune_gemm_backends": "ATEN,CUTLASS,TRITON",
        }
    ):
        with torch.no_grad():
            ep = torch.export.export(model, example)
            torch._inductor.aoti_compile_and_package(
                ep, package_path=package_path
            )

    # Extract wrapper.cpp from the .pt2 package (it's a zip).
    extract_dir = out_dir / "extracted"
    extract_dir.mkdir(exist_ok=True)
    with zipfile.ZipFile(package_path) as zf:
        cpp_members = [n for n in zf.namelist() if n.endswith(".cpp")]
        if not cpp_members:
            print(f"  [run {run_idx}] no .cpp in {package_path}; members: {zf.namelist()[:10]}")
            return {}
        # Largest .cpp is the wrapper
        cpp_members.sort(key=lambda n: zf.getinfo(n).file_size, reverse=True)
        wrapper_member = cpp_members[0]
        zf.extract(wrapper_member, extract_dir)
    wrapper_cpp = extract_dir / wrapper_member
    text = wrapper_cpp.read_text()
    counts = {
        "cuBLAS_addmm_out": len(re.findall(r"aoti_torch_cuda_addmm_out", text)),
        "cuBLAS_mm_out": len(re.findall(r"aoti_torch_cuda_mm_out", text)),
        "triton_tem_fused_addmm": len(re.findall(r"triton_tem_fused_addmm", text)),
        "triton_poi_fused": len(re.findall(r"triton_poi_fused", text)),
    }
    print(f"  [run {run_idx}] wrapper={wrapper_cpp.name} {counts}")
    return counts


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--runs", type=int, default=4)
    args = parser.parse_args()

    with tempfile.TemporaryDirectory(prefix="repro180811_") as td:
        workdir = Path(td)
        print(f"workdir={workdir}")
        counts = []
        for i in range(args.runs):
            print(f"=== Run {i} ===")
            c = run_one(i, workdir)
            counts.append(c)

        print()
        print("Summary per run:")
        for i, c in enumerate(counts):
            print(f"  run {i}: {c}")
        keys = list(counts[0].keys())
        deterministic = all(len({c[k] for c in counts}) == 1 for k in keys)
        if deterministic:
            print(f"DETERMINISTIC across {args.runs} runs.")
        else:
            print("NON-DETERMINISTIC across runs:")
            for k in keys:
                vals = [c[k] for c in counts]
                if len(set(vals)) > 1:
                    print(f"  {k}: {vals}")


if __name__ == "__main__":
    main()

```

**Comment by AllenFeiZZ** [comment]
Thanks for testing @yushangdi 

A few clarifications:

### 1. Non-determinism is not 100% reproducible

The non-deterministic behavior we observed was on **PyTorch 2.8.0 + Ada Lovelace (SM 8.9)**. It's possible that on H100 (SM 9.0) or on the latest nightly, the autotuner's benchmark noise characteristics are different enough to produce deterministic results. The non-determinism was the symptom that led us to discover this issue, but it's not the core problem.

### 2. Your results actually confirm the core issue

Looking at your output:

```
cuBLAS_addmm_out: 0, cuBLAS_mm_out: 50, triton_tem_fused_addmm: 0, triton_poi_fused: 68
```

This means on the latest nightly, the autotuner **deterministically** selects cuBLAS (`mm_out`) for all 50 GEMM ops, and `triton_tem_fused_addmm` is **0** — no fused Triton template kernels were selected at all. The fusion between `addmm` and `gelu` is still broken in every run.

### 3. The real question: is this a bug or by-design?

The non-determinism aside, the deeper question is about how `max_autotune` makes backend selection decisions:

- The autotuner benchmarks each GEMM **in isolation** (cuBLAS vs Triton template), and picks the locally fastest one.
- But selecting cuBLAS has a **global cost** that isn't measured: it **breaks fusion** with the subsequent activation (GeLU/ReLU). This means the fused `addmm + gelu` single-kernel launch gets split into `cuBLAS mm_out` + a separate `triton_poi_fused_gelu` kernel — an extra kernel launch and an extra global memory round-trip.
- For small GEMM (M=1), this fusion overhead can dominate actual compute time. In our production model, Nsight profiling showed the 50 unfused cuBLAS calls consumed **~6.1% of total GPU time** — time that was essentially zero when Triton fused templates were selected.

So the question is: **should the autotuner account for fusion benefits when comparing cuBLAS (unfusable) vs Triton templates (fusable)?** Currently it appears to make a locally optimal choice (cuBLAS may be marginally faster for the isolated GEMM), but this leads to a globally suboptimal result (broken fusion, more kernel launches, more memory traffic).

Is this the intended behavior, or is there a way to make the autotuner fusion-aware?

**Comment by eellison** [comment]
```
3. The real question: is this a bug or by-design?
The non-determinism aside, the deeper question is about how max_autotune makes backend selection decisions:

The autotuner benchmarks each GEMM in isolation (cuBLAS vs Triton template), and picks the locally fastest one.
But selecting cuBLAS has a global cost that isn't measured: it breaks fusion with the subsequent activation (GeLU/ReLU). This means the fused addmm + gelu single-kernel launch gets split into cuBLAS mm_out + a separate triton_poi_fused_gelu kernel — an extra kernel launch and an extra global memory round-trip.
```

We do actually model this. See: https://github.com/pytorch/pytorch/blob/446033e622b25fea5fb72c3b1c5d9f9da2a11ef3/torch/_inductor/ir.py#L5643-L5650. 

You can also try https://github.com/pytorch/pytorch/pull/179246 for more stable benchmarking when it lands. 

Other options:
- set `TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKENDS="TRITON"`


- we will be landing a per shape, config overriding eventually as well, cc @mlazos
